source/extensions/filters/listener/dynamic_modules/abi_impl.cc

```diff
@@ -0,0 +1,316 @@
+#include "source/common/protobuf/utility.h"
+#include "source/common/router/string_accessor_impl.h"
+#include "source/extensions/dynamic_modules/abi.h"
+#include "source/extensions/filters/listener/dynamic_modules/filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace DynamicModules {
+namespace ListenerFilters {
+
+extern "C" {
+
+bool envoy_dynamic_module_callback_listener_filter_get_buffer_chunk(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_envoy_buffer* chunk_out) {
+  auto* filter = static_cast<DynamicModuleListenerFilter*>(filter_envoy_ptr);
+  Network::ListenerFilterBuffer* buffer = filter->currentBuffer();
+
+  if (buffer == nullptr) {
+    chunk_out->ptr = nullptr;
+    chunk_out->length = 0;
+    return false;
+  }
+
+  auto raw_slice = buffer->rawSlice();
+  chunk_out->ptr = static_cast<char*>(const_cast<void*>(raw_slice.mem_));
+  chunk_out->length = raw_slice.len_;
+  return true;
+}
+
+bool envoy_dynamic_module_callback_listener_filter_drain_buffer(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr, size_t length) {
+  auto* filter = static_cast<DynamicModuleListenerFilter*>(filter_envoy_ptr);
+  Network::ListenerFilterBuffer* buffer = filter->currentBuffer();
+
+  if (buffer == nullptr || length == 0) {
+    return false;
+  }
+
+  buffer->drain(length);
+  return true;
+}
+
+void envoy_dynamic_module_callback_listener_filter_set_detected_transport_protocol(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_module_buffer protocol) {
+  auto* filter = static_cast<DynamicModuleListenerFilter*>(filter_envoy_ptr);
+  auto* callbacks = filter->callbacks();
+  if (callbacks != nullptr && protocol.ptr != nullptr && protocol.length > 0) {
+    callbacks->socket().setDetectedTransportProtocol(
+        absl::string_view(protocol.ptr, protocol.length));
+  }
+}
+
+void envoy_dynamic_module_callback_listener_filter_set_requested_server_name(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_module_buffer name) {
+  auto* filter = static_cast<DynamicModuleListenerFilter*>(filter_envoy_ptr);
+  auto* callbacks = filter->callbacks();
+  if (callbacks != nullptr && name.ptr != nullptr && name.length > 0) {
+    callbacks->socket().setRequestedServerName(absl::string_view(name.ptr, name.length));
+  }
+}
+
+void envoy_dynamic_module_callback_listener_filter_set_requested_application_protocols(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_module_buffer* protocols, size_t protocols_count) {
+  auto* filter = static_cast<DynamicModuleListenerFilter*>(filter_envoy_ptr);
+  auto* callbacks = filter->callbacks();
+  if (callbacks == nullptr || protocols == nullptr || protocols_count == 0) {
+    return;
+  }
+
+  std::vector<absl::string_view> protocol_list;
```

Comment:
i think we can use inline vector here since setRequestedApplicationProtocols takes a ref to a vector so we could likely avoid the vector allocation?